### ISSUE TITLE
Feature request: warn about confusing finally

### ISSUE THREAD

**Issue by mk-pmb** [douglascrockford/JSLint#192]
Example code adapted from eslint's [no-unsafe-finally](http://eslint.org/docs/rules/no-unsafe-finally). People might think this function would return 1, especially if the catch block becomes rather long.

```js
(function () {
    "use strict";
    try {
        return 1;
    } catch (ignore) {
        return 2;
    } finally {
        return 3;
    }
}());
```

**Comment by douglascrockford** [comment]
I don't understand. Are you demanding that finally not be allowed? Can you please show an example of how this causes problems?

**Comment by mk-pmb** [comment]
The problem is that above function looks like it will return early on and its result will be 1, but it's 3. Similar if I replace the first return with a throw, readers might expect a result of 2, but again it's 3. (Tested in Firefox console and Node.js.)

As far as I can tell, the confusion only concerns return statements in a finally block, and only if the try block or any catch block had a reachable return in them. People could assume that when a function encounters a return statement, execution of that function ends and it will return the value given in that return statement, as the name suggests.

Changing the return value later on may be a very effective shuriken, but people without ninja training should be warned. If you can think of any other tricks that allow code to slip in between a return statement and the real end of the function, they probably belong in the same bucket.

**Comment by douglascrockford** [comment]
here is only one catch clause. You may be thinking of Java.

Can you show a single example of how this actually causes a problem? What exactly are you proposing?

**Comment by mk-pmb** [comment]
Fortunately, I haven't yet encountered in-the-wild bugs of this type, so I can't say whether this "actually causes" a problem. I'll let you know if I encounter some.
My proposal aimed at a warning like "This `return` inside a `finally` block might replace the earlier value `return`ed from inside the `try` (or `catch`) block".
Reserching the issue, I found that it can actually be a lot more complex if there are nested `try`s and loops, so I can't make up an easy rule for when exactly to warn. I'll rather just add a warning for me to be especially careful when there's flow control in my `finally`s. Will work for me since I don't usually have them anyways, but I wouldn't want to force that on other users.

**Comment by karimsa** [comment]
The `return` statements inside the try and catch blocks are unreachable due to the presence of the finally. This can be a problem if the purpose of `finally` is misunderstood by the programmer.

For instance, one might think that `finally` can be used in place of `catch` when you're not utilizing the error being thrown (JSLint will complain about the unused error variable unless there is an explicit `ignore` directive being used). This is shown in the following code:

```javascript
var importantFn = function (file) {
  try {
    return fs.readFileSync(file, 'utf8');
  } finally {
    return null;
  }
};
```

One might think this returns `null` upon error but it actually returns `null` at all times.

**Comment by mk-pmb** [comment]
> one might think that finally can be used in place of catch

I think that would make for a more grave misunderstanding than just the one I mentioned, expecting a function to end on return. Especially since a lot of books and tutorials that claim to teach JS, do tell you that a return will end your function.

**Comment by karimsa** [comment]
I agree with that. I was simply trying to demonstrate a dangerous case of control flow statements in the `finally` block.

I think JSLint should at least generate a warning for this.

**Comment by douglascrockford** [comment]
Please try it now.

**Comment by mk-pmb** [comment]
> Unexpected 'return'.

Thanks!
